### PR TITLE
feat(plugin): RelationshipNotifier

### PR DESCRIPTION
**Pull request by lelboatz**
This plugin will send show a notice at the top of the screen (and a desktop notification if discord isn't focused) when a friend removes you or cancels their friend request to you and when you are removed from a group chat or a server.

This also fixes transitionTo not working (credit to the sun from the vencord discord server)

### CODE REVIEW

File: src/plugins/relationshipNotifier/utils.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    const currentChannels: Channel[] = DMStore.getSortedPrivateChannels();
+    for (const channel of currentChannels) {
+        if (channel.type !== 3) continue;
+        groups.set(channel.id, {
+            id: channel.id,
+            name: channel.name ? channel.name : channel.rawRecipients.map(r => r.username).join(", "),
+            iconURL: channel.icon ? `https://cdn.discordapp.com/channel-icons/${channel.id}/${channel.icon}.png` : undefined
+        });
+    }
+}
```
Comment (Vendicated) [inline src/plugins/relationshipNotifier/utils.ts:81]: you don't seem to persist these anywhere. Why not do that (DataStore is the best way to do that) so you can also notify for changes while Discord was closed?

File: src/plugins/relationshipNotifier/utils.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import { Notices } from "@api/index";
+import { findByPropsLazy } from "@webpack";
+import { NavigationRouter } from "@webpack/common";
+import { Channel, Guild } from "discord-types/general";
+
+import relationshipNotifier from "./index";
```
[the remark this answers is not in the document]
Comment (lelboatz) [reply src/plugins/relationshipNotifier/utils.ts]: fixed

File: src/plugins/relationshipNotifier/utils.ts (hunk 1 above)
Comment (lelboatz) [reply src/plugins/relationshipNotifier/utils.ts:141]: done

**Comment by lelboatz** [comment]
i fixed the eslint issues, can you run the workflow again?

File: src/webpack/common.tsx [hunk 3]
```diff
@@ -152,9 +153,9 @@ export const Clipboard = mapMangledModuleLazy('document.queryCommandEnabled("cop
     SUPPORTS_COPY: x => typeof x === "boolean",
 });
 
-export const NavigationRouter = mapMangledModuleLazy('"transitionToGuild', {
-    transitionTo: filters.byCode('"transitionTo '),
-    transitionToGuild: filters.byCode('"transitionToGuild'),
+export const NavigationRouter = mapMangledModuleLazy("Transitioning to external path", {
+    transitionTo: filters.byCode("Transitioning to external path"),
```
Comment (lelboatz) [inline src/webpack/common.tsx:157]: this breaks transitionTo

Comment (Vendicated) [reply src/webpack/common.tsx:157]: it's already fixed

Comment (lelboatz) [reply src/webpack/common.tsx:157]: oh ok cool

**Comment by Vendicated** [comment]
no idea why restore didnt work. In any case i fixed it now

File: src/plugins/relationshipNotifier/settings.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import { definePluginSettings } from "@api/settings";
+import { OptionType } from "@utils/types";
+
+const settings = definePluginSettings({
+    notifications: {
+        type: OptionType.BOOLEAN,
+        description: "Send a desktop notification discord is unfocused",
```
Comment (Vendicated) [inline src/plugins/relationshipNotifier/settings.ts:25]:
```suggestion
        description: "Also send a desktop notification when discord is unfocused",
```

File: src/plugins/relationshipNotifier/settings.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import { definePluginSettings } from "@api/settings";
+import { OptionType } from "@utils/types";
+
+const settings = definePluginSettings({
```
Comment (Vendicated) [inline src/plugins/relationshipNotifier/settings.ts:22]: why not just ```suggestion
export default definePluginSettings({
```

**Comment by Vendicated** [comment]
seems to work pretty well but the top banners are a really bad way to notify for this imo since they lack pfp capabilities and it's just ugh. I've been planning to implement neat in-app notifications for a while now

**Comment by lelboatz** [comment]
> seems to work pretty well but the top banners are a really bad way to notify for this imo since they lack pfp capabilities and it's just ugly. I've been planning to implement neat in-app notifications for a while now

i agree it would be nice to have an in-app notification system but the main reason i chose top banners was because people often miss in-app notifications if they have discord open on one screen while doing something else. top banners will stay on the screen until dismissed or discord is restarted. i would however be more than happy to switch to another notification system if it provided some sort of notification log and/or a way to keep them on screen until the user interacts with it.

**Comment by Vendicated** [comment]
Hi! I just merged Notifications. They don't currently have a way to not auto dismiss but you could add that if you wanted

**Comment by lelboatz** [comment]
> Hi! I just merged Notifications. They don't currently have a way to not auto dismiss but you could add that if you wanted

will add notifications to this soon, as for auto dismiss i'll figure something out for that

**Comment by lelboatz** [comment]
implemented the notification api. i also removed the setting which toggles desktop notifications while discord is unfocused since you added that in the api and added an option to show notices at the top of the screen as well (instead of making notifications not auto dismiss)

**Comment by Vendicated** [comment]
> i also removed the setting which toggles desktop notifications while discord is unfocused

what?

**Comment by Vendicated** [comment]
oh you were talking about the plugin's feature, i thought you removed it from the notificiations api

File: src/plugins/relationshipNotifier/utils.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+import { RelationshipStore, UserUtils } from "@webpack/common";
+import { Channel, Guild } from "discord-types/general";
+
+import settings from "./settings";
+import SimpleGroupChannel from "./types/SimpleGroupChannel";
+import SimpleGuild from "./types/SimpleGuild";
+
+const DMStore = findByPropsLazy("getSortedPrivateChannels");
+const GuildStore = findByPropsLazy("getGuilds", "getGuild");
+const Notifications = findByPropsLazy("showNotification", "requestPermission");
```
Comment (Vendicated) [inline src/plugins/relationshipNotifier/utils.ts:31]: unused?

File: src/plugins/relationshipNotifier/utils.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+import { findByPropsLazy } from "@webpack";
+import { RelationshipStore, UserUtils } from "@webpack/common";
+import { Channel, Guild } from "discord-types/general";
+
+import settings from "./settings";
+import SimpleGroupChannel from "./types/SimpleGroupChannel";
+import SimpleGuild from "./types/SimpleGuild";
+
+const DMStore = findByPropsLazy("getSortedPrivateChannels");
+const GuildStore = findByPropsLazy("getGuilds", "getGuild");
```
Comment (Vendicated) [inline src/plugins/relationshipNotifier/utils.ts:30]: this is in webpack commons!

File: src/plugins/relationshipNotifier/utils.ts [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+const groups = new Map<string, SimpleGroupChannel>();
+const friends: { friends: string[]; requests: string[] } = {
+    friends: [],
+    requests: []
+};
+
+export async function syncAndRunChecks() {
+    const oldGuilds: Map<string, SimpleGuild> | undefined = await DataStore.get("relationship-notifier-guilds");
+    const oldGroups: Map<string, SimpleGroupChannel> | undefined = await DataStore.get("relationship-notifier-groups");
+    const oldFriends: { friends: string[]; requests: string[] } | undefined = await DataStore.get("relationship-notifier-friends");
```
Comment (Vendicated) [inline src/plugins/relationshipNotifier/utils.ts:43]: you could Promise.all these
```ts
const [oldGuilds, oldGroups, oldFriends] = await Promise.all([
    DataStore.get("relationship-notifier-guilds"),
    DataStore.get("relationship-notifier-groups"),
    DataStore.get("relationship-notifier-friends")
]);
```

Comment (Vendicated) [reply src/plugins/relationshipNotifier/utils.ts:43]: actually there's DataStore.getMany

Comment (lelboatz) [reply src/plugins/relationshipNotifier/utils.ts:41]: how do i type the vars with DataStore.getMany

Comment (Vendicated) [reply src/plugins/relationshipNotifier/utils.ts:41]: Sorry I forgot to reply.

```ts
const [oldGuilds, oldGroups, oldFriends] = await Promise.all([
    DataStore.get("relationship-notifier-guilds"),
    DataStore.get("relationship-notifier-groups"),
    DataStore.get("relationship-notifier-friends")
]) as [Map<string, SimpleGuild>, Map<string, SimpleGroupChannel>, Record<"friends" | "requests", string[]>];
```

**Comment by lelboatz** [comment]
this is ready to be merged btw (unless someone wants to review some more)

**Comment by Vendicated** [comment]
Since the code wasn't really matching the style used by the rest of the project, I rewrote / cleaned up some parts

**Pull request merged**